[14.0] Add module edi_bank_statement_import

TeoGoddet opened the pull request:
This module plugs edi framework into account_statement_import from OCA/bank-statement-import
It allow to parse data retrieved trough the framework as bank statement.


It's heavily based on https://github.com/Noviat/account_ebics/blob/14.0/account_ebics/models/ebics_file.py and #451

TeoGoddet commented on edi_bank_statement_import/components/process.py:
```diff
@@ -0,0 +1,74 @@
+                    statement.button_post()
+                names.append(statement.name)
+    
+                # TODO : I don't know what to do when there is multiple statements, 
```
Hi !
Here it's possible to have mutiple statements in a single xml file, that mean multiple records for one exchange.
I suppose that it's possible to create child exchange but i could not find infos/example on how it should be done.

simahawk replied:
@TeoGoddet we've discussed this recently. IMO a good approach could be:

1. make exchange_record.exchange_file computed and write-able
2. if the record is a child and there's no value look at the parent

If you take action on this, please do it in a separated PR.

TeoGoddet replied:
#542

simahawk commented on edi_bank_statement_import/components/process.py:
```diff
@@ -0,0 +1,81 @@
+            )
+            return True
+
+        if len(result['statement_ids']) == 1:
```
pls split this `process` method into smaller methods

TeoGoddet replied:
done

simahawk commented on edi_bank_statement_import/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,7 @@
+This module allows to use account_bank_statement_import to parse edi exchange (input).
+This allow you ro retrieve bank stazement from a any edi exchange backend (webservice, sftp, local filesystem, ...)
+
+TODO: shall we add an exchange type example as demo?
```
yes pls :)

TeoGoddet replied:
I'm not really able to, but let the message (inspired from #451)

TeoGoddet replied:
Done, please check it

simahawk replied:
See my comment above

simahawk commented on edi_bank_statement_import/tests/edi_bank_statement_import_test.py:
```diff
@@ -0,0 +1,161 @@
+
+    def setUp(self):
+        super(TestImport, self).setUp()
+        bank = self.env["res.partner.bank"].create(
```
this code runs for every test. Please use `SavepointCase` + `setUpClass` and if you don't rely on tracking also use this line:
```
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```
You'll see this approach in edi_oca too.

TeoGoddet replied:
removed the tests

simahawk commented on edi_bank_statement_import/tests/edi_bank_statement_import_test.py:
```diff
@@ -0,0 +1,161 @@
+        )
+
+    def test_statement_import(self):
+        """Test correct creation of single statement."""
```
here you are testing the base wizard, this should be done in the base module not here unless you are testing a special behavior provided by this new module.

TeoGoddet replied:
removed the tests

simahawk commented on edi_bank_statement_import/tests/edi_bank_statement_import_test.py:
```diff
@@ -0,0 +1,161 @@
+            )
+
+    def test_zip_import(self):
+        """Test import of multiple statements from zip file."""
```
same.

TeoGoddet replied:
removed the tests

simahawk commented on edi_bank_statement_import/tests/edi_bank_statement_import_test.py:
```diff
@@ -0,0 +1,161 @@
+            )
+
+        self.assertTrue(all([st.line_ids for st in bank_st_record]))
+        self.assertEqual(bank_st_record[0].line_ids.mapped("sequence"), [1, 2, 3])
```
pls have a look at how tests are done in edi_oca and test the component.

TeoGoddet replied:
removed the tests

TeoGoddet committed:
[ADD] edi_bank_statement_import

This module plugs edi framework into account_statement_import from OCA/bank-statement-import
It allow to parse as bank statement data retrieved trough the framework.

TeoGoddet commented:
Hello @OCA/edi-maintainers, 
I'm sorry to bother you, but I don't understand why the test fails on another module. Does someone have an idea ?

TeoGoddet commented:
Hello @simahawk, 
I've added a demo and some tests (based on your work for sale order [75b81519f61aa385c3b412be4049aeeaf60736b1](https://github.com/OCA/edi/commit/75b81519f61aa385c3b412be4049aeeaf60736b1))

etobella reviewed (approved):
Makes sense to me.

etobella commented:
@simahawk As original reviewer, can you give us your opinions ? :pray:

simahawk commented on edi_bank_statement_import/components/process.py:
```diff
@@ -0,0 +1,92 @@
+            "notifications": [],
+        }
+
+        file_data = base64.b64decode(wiz.statement_file)
+        wiz.import_single_file(file_data, result)
+
+        return self._handle_result(result["statement_ids"], wiz.statement_filename)
+
+    def _handle_result(self, statement_ids, filename):
+        if not statement_ids:
+            self.exchange_record.sudo().message_post(
+                body=_(
+                    "You have already imported this file (%s) or this file "
+                    "only contains already imported transactions."
+                )
+                % filename
+            )
+            return True
+
+        if len(statement_ids) == 1:
+            self._handle_single_result(self.exchange_record, statement_ids[0])
+            return True
+        else:
+            for statement_id in statement_ids:
+
+                child_exchange = self.exchange_record.create(
+                    {
+                        "identifier": "%s_%s"
+                        % (self.exchange_record.identifier, statement_id),
+                        "type_id": self.exchange_record.type_id.id,
+                        "backend_id": self.exchange_record.backend_id.id,
+                        "edi_exchange_state": "input_processed",
+                        "parent_id": self.exchange_record.id,
+                    }
+                )
+                self._handle_single_result(child_exchange, statement_id)
+                self.exchange_record.sudo().message_post(
+                    body=_("Multiple Bank Statements created (See child exchanges)")
+                )
+            return True
```
Not a blocker, but nice to have:

```suggestion
        file_data = base64.b64decode(wiz.statement_file)
        wiz.import_single_file(file_data, result)
        
        if not statement_ids:
            return self._handle_already_imported()
        if len(statement_ids) == 1:
            return self._handle_single_result(self.exchange_record, statement_ids[0])
        return self._handle_result(result["statement_ids"], wiz.statement_filename)
   
    def _handle_already_imported(self, wiz):
            self.exchange_record.sudo().message_post(
                body=_(
                    "You have already imported this file (%s) or this file "
                    "only contains already imported transactions."
                )
                % wiz.statement_filename
            )
            return False
            
    def _handle_result(self, statement_ids, filename):
            for statement_id in statement_ids:

                child_exchange = self.exchange_record.create(
                    {
                        "identifier": "%s_%s"
                        % (self.exchange_record.identifier, statement_id),
                        "type_id": self.exchange_record.type_id.id,
                        "backend_id": self.exchange_record.backend_id.id,
                        "edi_exchange_state": "input_processed",
                        "parent_id": self.exchange_record.id,
                    }
                )
                self._handle_single_result(child_exchange, statement_id)
                self.exchange_record.sudo().message_post(
                    body=_("Multiple Bank Statements created (See child exchanges)")
                )
            return True
```

simahawk replied:
Well, the formatting is a bit messed up, but I think you get my point ;)

simahawk commented on edi_bank_statement_import/components/process.py:
```diff
@@ -0,0 +1,92 @@
+            "statement_filename": self.exchange_record.exchange_filename,
+            "statement_file": self.exchange_record._get_file_content(binary=False),
+        }
+        wiz = self.env[wiz_model].with_context(wiz_ctx).sudo().create(wiz_vals)
```
Recommendation for all the cases where you use `with_context`: never override the full ctx.
Always explode the keys to not override the rest:
```suggestion
        wiz = self.env[wiz_model].with_context(**wiz_ctx).sudo().create(wiz_vals)
```

simahawk commented on edi_bank_statement_import/data/data.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
Do you really think we need a specific backend type for this?
We should probably have a generic one into a new module `edi_bank`?
In any case, if you keep it, please move it to its own file and make it noupdate.

simahawk commented on edi_bank_statement_import/data/data.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="name">Bank Statement Backend Type</field>
+        <field name="code">bank_statement_backend_type</field>
+    </record>
+    <record id="edi_exchange_type" model="edi.exchange.type">
```
Creating an exchange type by default w/out noupdate is dangerous.

BTW personally I'm thinking about removing default exchange types from the modules I maintain because it can be dangerous too due to their high level of configuration.

I tend to say that you should leave this to the specific setup of the instance and move this to demo data.
Just a recommendation tho. I'd like to hear your feedback and @etobella 's too.

simahawk commented on edi_bank_statement_import/data/data.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="edi_exchange_backend" model="edi.backend.type">
+        <field name="name">Bank Statement Backend Type</field>
+        <field name="code">bank_statement_backend_type</field>
```
```suggestion
        <field name="name">Bank Statement</field>
        <field name="code">bank_statement</field>
```

simahawk reviewed (changes requested):
Thanks for your work, I left some more remarks.
